### PR TITLE
Mempool support for transaction expiration

### LINKED ISSUE #2622: Flush expired transactions from mempool

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

Transactions that are expired should be removed from the mempool.

Closes #2622

### Specifications

https://zips.z.cash/zip-0203

## Solution

In zebra each transaction has a `expiry_height()` method that returns an `Option<Height>`. If this is `None` then there is no expiration limit and transaction will remain in the mempool until they are mined (or flushed by other means outside the scope of this PR).
If the method returns a `Height` then this must be compared with the current tip to decide if the transaction should be removed form the mempool.

It was discussed in discord briefly that the mempool `poll_ready()` can be a good place to do this so this PR implement it there however other options can be tried. 

## Review

This is currently a draft and it is untested. The reason for that is that there is a lot going on with mempool tests in open PRs  (https://github.com/ZcashFoundation/zebra/pull/2769, https://github.com/ZcashFoundation/zebra/pull/2770, https://github.com/ZcashFoundation/zebra/pull/2771, https://github.com/ZcashFoundation/zebra/pull/2748, etc) that i prefer those to settle first before adding more code that will conflict.

Basically, it will be good to see if this is the way to go for this functionality, adding tests should not be hard but a must for this. I think @conradoplg  and/or @jvff can take a look if they have the time.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebrad/src/components/mempool.rs [hunk 1]
```diff
@@ -82,6 +82,9 @@ pub struct Mempool {
     /// Allows checking if we are near the tip to enable/disable the mempool.
     #[allow(dead_code)]
     sync_status: SyncStatus,
+
+    /// Access the state of the current blockchain.
+    state: State,
```
Comment (oxarbitrage) [inline zebrad/src/components/mempool.rs]: State is needed to get the blockchain tip height.

File: zebrad/src/components/mempool.rs [hunk 2]
```diff
@@ -141,6 +146,24 @@ impl Service<Request> for Mempool {
...
+        let _ = tokio::task::spawn(async move {
+            let tip_height = match state
+                .ready_and()
+                .await
+                .unwrap()
+                .call(zs::Request::Tip)
+                .await
+            {
+                Ok(zs::Response::Tip(Some((height, _hash)))) => height,
+                _ => panic!("A valid tip height is needed here"),
```
Comment (oxarbitrage) [inline zebrad/src/components/mempool.rs]: Some of the existing tests will fail for this, i think they need to be fixed as having a tip at this point should be a must.

File: zebrad/src/components/mempool.rs [hunk 3]
```diff
@@ -175,3 +198,22 @@ impl Service<Request> for Mempool {
...
+    let transactions = storage.transactions(ids);
+
+    let _ = transactions
+        .iter()
+        .filter(|t| t.transaction.expiry_height().is_some())
+        .map(|t| {
+            if tip_height > t.transaction.expiry_height().unwrap() {
+                storage.remove(&t.id);
+            }
+        });
```
Comment (oxarbitrage) [inline zebrad/src/components/mempool.rs]: Probably this can be done a bit better. It is also untested code.

Comment (teor2345) [reply zebrad/src/components/mempool.rs]: I think this code looks good, it's clear, and it does what it's supposed to do.

But there's a missing consensus rule:
> No limit: To set no limit on transactions (so that they do not expire), nExpiryHeight should be set to 0.

https://zips.z.cash/zip-0203#specification

We probably want to handle zeroes in `Transaction::expiry_height`, because this consensus rule applies to block transaction validation as well:
https://github.com/ZcashFoundation/zebra/blob/957e12e4ca28438ca2c64ca4a1870210f2b3e9b7/zebra-chain/src/transaction.rs#L294-L295

(This isn't a bug in this PR, it's something we wanted to fix for block validation in #2387, but we cut that ticket from our NU5 scope.)

File: zebrad/src/components/mempool.rs (hunk 2 above)
Comment (teor2345) [reply zebrad/src/components/mempool.rs]: Are you sure there will always be at least one verified block in the state?
Where is that checked?

If we're not checking, then Zebra with panic when a peer sends an inbound mempool request and:
* Zebra has initialized its network,
* but Zebra's state has no verified blocks.

(There's a `poll_ready` call inside each `mempool.oneshot(_)`.)

https://github.com/ZcashFoundation/zebra/blob/957e12e4ca28438ca2c64ca4a1870210f2b3e9b7/zebrad/src/components/inbound.rs#L298-L301

File: zebrad/src/components/mempool.rs [hunk 4]
```diff
@@ -141,6 +146,24 @@ impl Service<Request> for Mempool {
                 let _ = self.storage.insert(tx);
             }
         }
+
+        let mut storage = self.storage.clone();
+        let mut state = self.state.clone();
+
+        let _ = tokio::task::spawn(async move {
```
Comment (teor2345) [inline zebrad/src/components/mempool.rs]: This spawned task is great, it makes sure we won't deadlock if:
* the state buffer is full,
* and the state is waiting on extra blocks to verify a block,
* but the inbound service is waiting on mempool readiness before it can process the next block.

(The deadlock would time out after a few minutes anyway. But it would still impact performance and usability.)

File: zebrad/src/components/mempool.rs (hunk 1 above)
Comment (teor2345) [reply zebrad/src/components/mempool.rs]: Optional Minor Performance: `LatestChainTip::best_tip_height` checks a watch channel, which is more efficient than spawning an async task to run a state service request future.

(And the code is a bit simpler, too.)

https://github.com/ZcashFoundation/zebra/blob/44ac06775bf6bb77dea06d53a6fc5e04a79baf31/zebra-state/src/service/chain_tip.rs#L207-L208

There's a `LatestChainTip` in the `start` method:
https://github.com/ZcashFoundation/zebra/blob/957e12e4ca28438ca2c64ca4a1870210f2b3e9b7/zebrad/src/commands/start.rs#L58

I'm happy with whatever you decide here, I just wanted to make sure you knew there was another way to do it.

File: zebrad/src/components/mempool.rs (hunk 4 above)
Comment (teor2345) [reply zebrad/src/components/mempool.rs]: If you decide to switch to `LatestChainTip::best_tip_height`, you won't need to spawn a task, because `LatestChainTip` stores the latest tip height every time the state changes.

**Review by teor2345 — COMMENTED** [review]
This PR looks good.

There's one potential panic, and one consensus rule that we need to implement in this PR, because we cut it from our NU5 scope.

I'm looking forward to seeing some tests, when the testing framework all comes together.

File: zebrad/src/components/mempool.rs (hunk 1 above)
Comment (oxarbitrage) [reply zebrad/src/components/mempool.rs]: This is a LOT better and it is very easy/convenient to use in this case, thank you very much. I made the changes to use `LatestChainTip` at https://github.com/ZcashFoundation/zebra/pull/2774/commits/fc47b1403f75e8030e07d3e94c696af50f436f9c

File: zebrad/src/components/mempool.rs (hunk 2 above)
Comment (oxarbitrage) [reply zebrad/src/components/mempool.rs]: With https://github.com/ZcashFoundation/zebra/pull/2774/commits/fc47b1403f75e8030e07d3e94c696af50f436f9c the panic is now gone as we ignore the code if we have no block tip from `best_tip_height()`. This looks correct to me however we can add an `else` arm if we need to handle this.

File: zebrad/src/components/mempool.rs (hunk 3 above)
Comment (oxarbitrage) [reply zebrad/src/components/mempool.rs]: Yea, i assumed we already had code for this but clearly we don't. I think we should return `None` if `expiry_height` field is `0`. The code should be simple but i agree we can work on this in a separated PR in the context of https://github.com/ZcashFoundation/zebra/issues/2387

File: zebrad/src/components/mempool.rs (hunk 2 above)
Comment (teor2345) [reply zebrad/src/components/mempool.rs]: I don't think we need to remove any transactions when the state is empty.
Can you double-check my reasoning here?

* transactions with expiry height `0` never expire, this is a consensus rule
* the lowest valid expiry height is `1`
* the lowest tip height that can expire transactions is `2`, because transactions expire when the tip is greater
* if the state is empty (`tip_height == None`), it hasn't reached height `2` yet, so there are no expired transactions

So this code looks correct. But let's test `remove_expired_transactions` with expiry and tip heights of `0`, `1`, and `2`, to prevent future bugs.

File: zebrad/src/components/mempool.rs (hunk 3 above)
Comment (teor2345) [reply zebrad/src/components/mempool.rs]: >  I think we should return `None` if `expiry_height` field is `0`. The code should be simple but i agree we can work on this in a separated PR in the context of #2387

Sounds good.

I think handling `0` correctly might be a blocker for merging this PR.
Currently, this PR expires transactions with expiry height `0` as soon as block `1` is committed to the state.
But expiry height `0` should never expire.

Did you want to write the PR to handle `0`?

Comment (oxarbitrage) [reply zebrad/src/components/mempool.rs]: Yea, will do that.

File: zebrad/src/components/mempool.rs [hunk 5]
```diff
@@ -141,6 +148,14 @@ impl Service<Request> for Mempool {
                 let _ = self.storage.insert(tx);
             }
         }
+
+        let mut storage = self.storage.clone();
+        if let Some(tip_height) = self.latest_chain_tip.best_tip_height() {
+            let _ = tokio::task::spawn(async move {
+                remove_expired_transactions(&mut storage, tip_height).await
+            });
```
Comment (teor2345) [inline zebrad/src/components/mempool.rs]: Performance & Simplification Nitpick: Does `remove_expired_transactions` need to be async?

```suggestion
            remove_expired_transactions(&mut storage, tip_height);
```

File: zebrad/src/components/mempool.rs (hunk 4 above)
Comment (oxarbitrage) [reply zebrad/src/components/mempool.rs]: Sorry, i missed this one, will remove the task spawn.

File: zebrad/src/components/mempool.rs (hunk 3 above)
Comment (teor2345) [reply zebrad/src/components/mempool.rs]: Done in #2782.

File: zebrad/src/components/mempool.rs [hunk 6]
```diff
@@ -141,6 +148,11 @@ impl Service<Request> for Mempool {
                 let _ = self.storage.insert(tx);
             }
         }
+
+        if let Some(tip_height) = self.latest_chain_tip.best_tip_height() {
+            let _ = remove_expired_transactions(&mut self.storage, tip_height);
```
Comment (teor2345) [inline zebrad/src/components/mempool.rs]: Nitpick: there's no return value here any more.

```suggestion
            remove_expired_transactions(&mut self.storage, tip_height);
```

**Review by teor2345 — COMMENTED** [review]
This is good to go once it's been rebased on `main`, and it has some tests.
(With or without the cleanup I just suggested.)

File: zebrad/src/components/mempool.rs [hunk 7]
```diff
@@ -188,3 +198,19 @@ impl Service<Request> for Mempool {
         }
     }
 }
+
+fn remove_expired_transactions(
```
Comment (teor2345) [inline zebrad/src/components/mempool.rs:340]: Can you document this method, and the consensus rule it implements?

See:
https://zips.z.cash/zip-0203#specification

**Review by teor2345 — COMMENTED** [review]
This looks great, thanks for the test!

We're just missing documentation for the expiry height consensus rule.

File: zebrad/src/components/inbound/tests.rs [hunk 8]
```diff
@@ -171,13 +171,129 @@ async fn mempool_advertise_transaction_ids() -> Result<(), crate::BoxError> {
...
+        .zcash_deserialize_into()
+        .unwrap();
+    state_service
+        .clone()
+        .oneshot(zebra_state::Request::CommitFinalizedBlock(
+            block_one.clone().into(),
+        ))
+        .await
+        .unwrap();
+
```
Comment (teor2345) [inline zebrad/src/components/inbound/tests.rs:237]: Missing test cases:
- Can you check that `tx1` isn't expired after `block_one`? (it should only expire after block 2)
- Can you check that a transaction with no expiry height never expires?

I think that covers the consensus rules:
> Transactions will have a new field, nExpiryHeight, which will set the block height after which transactions will be removed from the mempool if they have not been mined.
> No limit: To set no limit on transactions (so that they do not expire), nExpiryHeight should be set to 0.

**Review by teor2345 — COMMENTED** [review]
I think we might need more tests, because we're going to refactor the code in this PR.
(It conflicts with other open PRs.)

File: zebrad/src/components/inbound/tests.rs (hunk 8 above)
Comment (oxarbitrage) [reply zebrad/src/components/inbound/tests.rs:237]: https://github.com/ZcashFoundation/zebra/pull/2774/commits/d39d84626be1402d4167bcf1ad9fa4d513321922

File: zebrad/src/components/mempool.rs (hunk 7 above)
Comment (oxarbitrage) [reply zebrad/src/components/mempool.rs:340]: https://github.com/ZcashFoundation/zebra/pull/2774/commits/532e3067fd47b38c82edca0188272ba84afd8576

**Review by teor2345 — DISMISSED** [review]
This is good to go once we resolve conflicts.

File: zebrad/src/components/mempool.rs [hunk 9]
```diff
@@ -322,3 +333,22 @@ impl Service<Request> for Mempool {
...
+    storage: &mut storage::Storage,
+    tip_height: zebra_chain::block::Height,
+) {
+    let ids = storage.tx_ids().iter().copied().collect();
+    let transactions = storage.transactions(ids);
+
+    for t in transactions {
+        if let Some(expiry_height) = t.transaction.expiry_height() {
+            if tip_height > expiry_height {
```
Comment (conradoplg) [inline zebrad/src/components/mempool.rs]: I _think_ this should be `>=`

> If the expiry is at block height  N, then the transaction must be included in block N or earlier. Block N+1 will be too late, and the transaction will be removed from the mempool.

If `expiry_height` is N, then the transaction must be included in block N or earlier. But if `tip_height` is also N then the tx can't be included in it since it was just mined.

```suggestion
            if tip_height >= expiry_height {
```

Comment (oxarbitrage) [reply zebrad/src/components/mempool.rs]: Thank you for that one, i changed this functionality at https://github.com/ZcashFoundation/zebra/pull/2774/commits/f7d8228783353e95cb49802e2327074cfc2bd8cb

**Review by conradoplg — APPROVED** [review]

**Comment by upbqdn** [comment]
I looked at the code, and it looks good to me too.
